rbac/management/group/relation_api_dual_write_group_handler.py

```diff
@@ -0,0 +1,142 @@
+
+from rbac.env import ENVIRONMENT
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class DualWriteException(Exception):
+    """DualWrite exception."""
+
+    pass
+
+
+class RelationApiDualWriteGroupHandler:
+    """Class to handle Dual Write API related operations."""
+
+    def __init__(self, group, event_type, principal_usernames, service_account_client_ids):
+        """Initialize RelationApiDualWriteGroupHandler."""
+        if not self.replication_enabled():
+            return
+        try:
+            self.group_relations = []
+            self.current_group_relations = []
+            self.principal_usernames = principal_usernames
+            self.service_account_client_ids = service_account_client_ids
+            self.group = group
+            self.event_type = event_type
+        except Exception as e:
+            raise DualWriteException(e)
+
+    def replication_enabled(self):
+        """Check whether replication enabled."""
+        return ENVIRONMENT.get_value("REPLICATION_TO_RELATION_ENABLED", default=False, cast=bool)
+
+    def get_current_group_relations(self):
+        """Get current groups relations."""
+        return self.current_group_relations
+
+    def generate_replication_event_to_outbox(self, group):
+        """Generate replication event to outbox table."""
+        if not self.replication_enabled():
+            return
+        self.group = group
+        self._generate_relations_for_group()
+        return self.save_replication_event_to_outbox()
+
+    def save_replication_event_to_outbox(self):
+        """Generate and store replication event to outbox table."""
+        if not self.replication_enabled():
+            return {}
+        try:
+            replication_event = self._build_replication_event()
+            self._save_replication_event(replication_event)
+        except Exception as e:
+            raise DualWriteException(e)
+        return replication_event
+
+    def _generate_relations_for_group(self):
+        """Generate relations a group."""
+        if not self.replication_enabled():
+            return []
+        try:
+            logger.info("[Dual Write] Generate new relations from group(%s): '%s'", self.group.uuid, self.group.name)
+
+            relations = []
+            principal_usernames = [item["username"] for item in self.principal_usernames]
+            service_account_client_ids = [item["clientId"] for item in self.service_account_client_ids]
+
+            filtered_principal_and_service_accounts_uuids = self.group.principals.filter(
+                Q(username__in=principal_usernames) | Q(service_account_id__in=service_account_client_ids)
+            ).values_list("uuid", flat=True)
+
+            for principal_uuid in filtered_principal_and_service_accounts_uuids:
+                relations.append(
+                    create_relationship("group", str(self.group.uuid), "user", str(principal_uuid), "member")
+                )
```

Comment:
May I ask why the additional query here?

I'm suspicious it would cause race conditions (and require us to lock the group) although I haven't proven that yet. But regardless of that, it maybe seems unnecessary?

If another tx removed the users from the group, that tx should commit and write its own replication event.

I'm hoping (perhaps naively) that group principal add/remove can be very simple, essentially 3 sql statements:

For a req to add principals

1. insert principals & insert group fks from principals in request
2. insert outbox event w/ principals in request
3. commit

And likewise for remove.

Reply:
Group is locked but query may be seems unnecessary. It was about what id of principal and service account we will have in relation. 

Request payload contains username of principal and clientID service account. 

so I will keep UUID as id in relation - eventually we can change it later.

Anyway any IDs could be passed as they are available from additions - it just need to be extracted and passed here.

I will remove query and pass UUID it here.